File: r/devScript_MAIN.R
# devScript_MAIN.R — Comprehensive walkthrough of socialchoicelab R package
# ============================================================================
# This script demonstrates user-facing functions across all major categories.
# Run in RStudio with devtools::load_all() or library(socialchoicelab).
#
# Scenario: Laing-Olmsted-Bear (A_Group_05)
#   Space:  0–200 × 0–200
#   Rule:   Simple majority (0.5)
#
# After any change to the R package source, run at the console:
   devtools::document()   # regenerate NAMESPACE + man pages
   devtools::install()    # reinstall the package
library(socialchoicelab)

# ===========================================================================
# 1. SETUP: Voter and alternative data in 2D policy space
# ===========================================================================

cat("\n=== 1. SETUP: Voter and Alternative Data (Laing-Olmsted-Bear) ===\n")

# Load the built-in scenario
scenario <- load_scenario("laing_olmsted_bear")

voters <- scenario$voters
sq     <- scenario$status_quo
alts   <- sq  # No other alternatives in this scenario

cat("Scenario:", scenario$name, "\n")
cat("Players (flat [x0, y0, x1, y1, ...]):\n", voters, "\n")
cat("Status quo:", sq, "\n")

# ===========================================================================
# 2. DISTANCE & UTILITY FUNCTIONS
# ===========================================================================

cat("\n=== 2. DISTANCE & UTILITY ===\n")

# Distance under different metrics
d_euclidean <- calculate_distance(c(0, 0), c(3, 4))
d_manhattan <- calculate_distance(c(0, 0), c(3, 4), make_dist_config("manhattan"))

cat("Distance from origin to (3,4):\n")
cat("  Euclidean:", d_euclidean, "\n")
cat("  Manhattan:", d_manhattan, "\n")

# Convert distance to utility
utility_at_2 <- distance_to_utility(2.0)
utility_at_5 <- distance_to_utility(5.0)

cat("Utility (linear loss):\n")
cat("  At distance 2:", utility_at_2, "\n")
cat("  At distance 5:", utility_at_5, "\n")

# Normalize utility to [0, 1]
utility_norm <- normalize_utility(utility_at_2, max_distance = 5)
cat("Normalized utility (max_distance=5):", utility_norm, "\n")

# ===========================================================================
# 3. MAJORITY PREFERENCE & PAIRWISE COMPARISONS
# ===========================================================================

cat("\n=== 3. MAJORITY PREFERENCE ===\n")

# Does a simple majority prefer each test point over the status quo?
# Test point A: slight shift toward Player 3 (upper-right quadrant)
# Test point B: slight shift toward Player 5 (lower-left quadrant)
a_beats_sq <- majority_prefers_2d(c(110.0, 110.0), sq, voters)
b_beats_sq <- majority_prefers_2d(c( 60.0,  60.0), sq, voters)

cat("Majority preference (5 voters):\n")
cat("  (110, 110) beats SQ:", a_beats_sq, "\n")
cat("  ( 60,  60) beats SQ:", b_beats_sq, "\n")

# Pairwise comparison matrix
pairwise_mat <- pairwise_matrix_2d(alts, voters)
cat("\nPairwise matrix (1=WIN, 0=TIE, -1=LOSS):\n")
print(pairwise_mat)

# ===========================================================================
# 4. GEOMETRY: COPELAND, CONDORCET, UNCOVERED SET
# ===========================================================================

cat("\n=== 4. GEOMETRY: COPELAND & CONDORCET ===\n")

# Copeland scores (wins minus losses in pairwise comparisons)
copeland <- copeland_scores_2d(alts, voters)
cat("Copeland scores:\n")
print(copeland)

# Copeland winner (ties broken by smallest index)
copeland_winner <- copeland_winner_2d(alts, voters)
cat("Copeland winner (1-based):", copeland_winner, "\n")

# Does a Condorcet winner exist?
has_cw <- has_condorcet_winner_2d(alts, voters)
cw <- condorcet_winner_2d(alts, voters)
cat("Has Condorcet winner:", has_cw, "\n")
cat("Condorcet winner (1-based):", cw, "\n")

# Uncovered set (Miller covering relation)
uncovered <- uncovered_set_2d(alts, voters)
cat("Uncovered set (1-based indices):", uncovered, "\n")

# ===========================================================================
# 5. WINSET: MAJORITY WINSET OF STATUS QUO
# ===========================================================================

cat("\n=== 5. WINSET ===\n")

# Create a winset: all policies that beat SQ under majority rule
winset <- winset_2d(sq, voters)

cat("Winset of SQ at (0, 0):\n")
cat("  Empty:", winset$is_empty(), "\n")

if (!winset$is_empty()) {
  bbox <- winset$bbox()
  cat("  Bounding box: [", bbox$min_x, ",", bbox$max_x, "] × [",
      bbox$min_y, ",", bbox$max_y, "]\n", sep="")

  cat("  Contains (110, 80):", winset$contains(110, 80), "\n")
  cat("  Contains (150, 150):", winset$contains(150, 150), "\n")
}

# ===========================================================================
# 6. ORDINAL PROFILES & VOTING RULES
# ===========================================================================

cat("\n=== 6. ORDINAL PROFILES & VOTING RULES ===\n")

# Build a preference profile from spatial locations
# Voters rank alternatives by distance (closest first)
profile <- profile_build_spatial(alts, voters)

cat("Profile dimensions:", profile$dims()$n_voters, "voters,",
    profile$dims()$n_alts, "alternatives\n")

cat("\nRankings (1-based indices, 1=best):\n")
print(profile$export())

# Plurality
plurality <- plurality_scores(profile)
plurality_winner <- plurality_one_winner(profile)
cat("\nPlurality scores:\n")
print(plurality)
cat("Plurality winner (1-based):", plurality_winner, "\n")

# Borda count
borda <- borda_scores(profile)
borda_winner <- borda_one_winner(profile)
borda_ranking <- borda_ranking(profile)
cat("\nBorda scores:\n")
print(borda)
cat("Borda winner (1-based):", borda_winner, "\n")
cat("Borda ranking (best first):", borda_ranking, "\n")

# Anti-plurality
antiplurality <- antiplurality_scores(profile)
antiplurality_winner <- antiplurality_one_winner(profile)
cat("\nAnti-plurality scores:\n")
print(antiplurality)
cat("Anti-plurality winner (1-based):", antiplurality_winner, "\n")

# Approval voting (top-k): k capped at n_alts so this works for any scenario
k_approval <- min(2L, profile$dims()$n_alts)
approval_topk <- approval_scores_topk(profile, k = k_approval)
approval_topk_winners <- approval_all_winners_topk(profile, k = k_approval)
cat(sprintf("\nApproval voting (top-%d):\n", k_approval))
cat("  Scores:", approval_topk, "\n")
cat("  Winners (1-based):", approval_topk_winners, "\n")

# ===========================================================================
# 7. SOCIAL RANKINGS & PROPERTIES
# ===========================================================================

cat("\n=== 7. SOCIAL RANKINGS & PROPERTIES ===\n")

# Rank by scores
scores <- c(1.5, 0.8, 2.1)
ranking <- rank_by_scores(scores)
cat("Rank by scores [1.5, 0.8, 2.1]:\n")
cat("  Best-to-worst (1-based):", ranking, "\n")

# Pareto set (Pareto-efficient alternatives)
pareto <- pareto_set(profile)
cat("\nPareto set (1-based):", pareto, "\n")

for (i in seq_len(profile$dims()$n_alts)) {
  is_pareto <- is_pareto_efficient(profile, i)
  cat("  Alt", i, "is Pareto-efficient:", is_pareto, "\n")
}

# Condorcet & majority consistency
has_cw_prof <- has_condorcet_winner_profile(profile)
cw_prof <- condorcet_winner_profile(profile)
cat("\nProfile has Condorcet winner:", has_cw_prof, "\n")
cat("Condorcet winner (profile, 1-based):", cw_prof, "\n")

if (!is.na(cw_prof)) {
  is_majority_cons <- is_majority_consistent(profile, cw_prof)
  is_condorcet_cons <- is_condorcet_consistent(profile, cw_prof)
  cat("  Is majority-consistent:", is_majority_cons, "\n")
  cat("  Is Condorcet-consistent:", is_condorcet_cons, "\n")
}

# ===========================================================================
# 8. STREAM MANAGER & REPRODUCIBILITY
# ===========================================================================

cat("\n=== 8. STREAM MANAGER & REPRODUCIBILITY ===\n")

# Create a stream manager with a fixed seed
sm <- stream_manager(master_seed = 42, stream_names = c("draws", "ties"))

# Draw random values (reproducible)
draw1 <- sm$uniform_real("draws", 0, 1)
draw2 <- sm$uniform_real("draws", 0, 1)

cat("Draws from stream 'draws':\n")
cat("  Draw 1:", draw1, "\n")
cat("  Draw 2:", draw2, "\n")

# Reset to get the same values
sm$reset_all(42)
draw1_again <- sm$uniform_real("draws", 0, 1)
draw2_again <- sm$uniform_real("draws", 0, 1)

cat("After reset (master_seed=42):\n")
cat("  Draw 1:", draw1_again, "(equal:", draw1 == draw1_again, ")\n")
cat("  Draw 2:", draw2_again, "(equal:", draw2 == draw2_again, ")\n")

# Generate a stochastic profile (impartial culture)
sm_ic <- stream_manager(master_seed = 99, stream_names = "ic")
profile_ic <- profile_impartial_culture(n_voters = 10L, n_alts = 4L,
                                        stream_manager = sm_ic, stream_name = "ic")
cat("\nImpartial culture profile (10 voters, 4 alts):\n")
cat("  Dims:", profile_ic$dims()$n_voters, "×", profile_ic$dims()$n_alts, "\n")
cat("  Voter 1 ranking:", profile_ic$get_ranking(1L), "\n")

# ===========================================================================
# 9. CONFIGURATION FUNCTIONS
# ===========================================================================

cat("\n=== 9. CONFIGURATION ===\n")

# Distance configs
dist_euc <- make_dist_config("euclidean", weights = c(1.0, 1.0))
dist_l1 <- make_dist_config("manhattan")
dist_weighted <- make_dist_config("euclidean", weights = c(2.0, 0.5))

cat("Distance configs:\n")
cat("  Euclidean (equal weights):\n")
cat("    - distance_type:", dist_euc$distance_type, "\n")
cat("    - weights:", paste(dist_euc$weights, collapse=", "), "\n")

# Loss configs
loss_linear <- make_loss_config("linear")
loss_quad <- make_loss_config("quadratic")
loss_gaussian <- make_loss_config("gaussian", max_loss = 2.0, steepness = 0.5)

cat("\n  Loss configs:\n")
cat("    - Linear:", loss_linear$loss_type, "\n")
cat("    - Quadratic:", loss_quad$loss_type, "\n")
cat("    - Gaussian (max_loss=2, steepness=0.5):", loss_gaussian$loss_type, "\n")

# ===========================================================================
# 10. API VERSION
# ===========================================================================

cat("\n=== 10. API VERSION ===\n")

version <- scs_version()
cat("C API version:", version$major, ".", version$minor, ".", version$patch, "\n")

# ===========================================================================
# 10. VISUALIZATION (requires plotly — install.packages("plotly") once)
# ===========================================================================

cat("\n=== 10. VISUALIZATION ===\n")

hull <- convex_hull_2d(voters)
ws   <- winset_2d(sq, voters)
bnd  <- uncovered_set_boundary_2d(voters, grid_resolution = 10L)

# Base plot: players + status quo only (no additional alternatives in this scenario)
fig <- plot_spatial_voting(
  voters,
  sq          = sq,
  voter_names = scenario$voter_names,
  dim_names   = scenario$dim_names,
  title       = scenario$name
)

# Add layers
fig <- layer_winset(fig, ws)
fig <- layer_convex_hull(fig, hull)
fig <- finalize_plot(fig)

# Display (opens in RStudio Viewer / browser)
print(fig)
cat("Plot displayed in Viewer. Hover over points for details.\n")

# ===========================================================================
# END OF SCRIPT
# ===========================================================================

cat("\n=== END ===\n")
cat("All examples completed successfully!\n")
cat("For more detail, see: ?function_name or vignette('spatial_voting')\n")

File: python/devScript_MAIN.py
"""devScript_MAIN.py — Comprehensive walkthrough of socialchoicelab Python package
================================================================================
This script demonstrates user-facing functions across all major categories.
Run with the project .venv:

    .venv/bin/python python/devScript_MAIN.py

Or activate the venv first:

    source .venv/bin/activate
    python python/devScript_MAIN.py

Scenario: Laing-Olmsted-Bear (A_Group_05)
  Space:  0–200 × 0–200
  Rule:   Simple majority (0.5)

The SCS_LIB_PATH environment variable must point to the compiled library:

    export SCS_LIB_PATH=/path/to/socialchoicelab/build
"""

import os
import tempfile
import webbrowser
from pathlib import Path

import numpy as np
import socialchoicelab as scl

# ===========================================================================
# 1. SETUP: Voter and alternative data in 2D policy space
# ===========================================================================

print("\n=== 1. SETUP: Voter and Alternative Data (Laing-Olmsted-Bear) ===")

# Load the built-in scenario
scenario = scl.load_scenario("laing_olmsted_bear")

voters = scenario["voters"]
sq = scenario["status_quo"]
alts = sq  # No other alternatives in this scenario

print(f"Scenario: {scenario['name']}")
print("Players (flat [x0, y0, x1, y1, ...]):", voters)
print("Status quo:", sq)

# ===========================================================================
# 2. DISTANCE & UTILITY FUNCTIONS
# ===========================================================================

print("\n=== 2. DISTANCE & UTILITY ===")

d_euclidean = scl.calculate_distance([0.0, 0.0], [3.0, 4.0])
d_manhattan = scl.calculate_distance([0.0, 0.0], [3.0, 4.0],
                                     scl.make_dist_config("manhattan"))

print("Distance from origin to (3, 4):")
print(f"  Euclidean: {d_euclidean}")
print(f"  Manhattan: {d_manhattan}")

utility_at_2 = scl.distance_to_utility(2.0)
utility_at_5 = scl.distance_to_utility(5.0)

print("Utility (linear loss):")
print(f"  At distance 2: {utility_at_2}")
print(f"  At distance 5: {utility_at_5}")

utility_norm = scl.normalize_utility(utility_at_2, max_distance=5.0)
print(f"Normalized utility (max_distance=5): {utility_norm}")

# ===========================================================================
# 3. MAJORITY PREFERENCE & PAIRWISE COMPARISONS
# ===========================================================================

print("\n=== 3. MAJORITY PREFERENCE ===")

a_beats_sq = scl.majority_prefers_2d(110.0, 110.0, sq[0], sq[1], voters)
b_beats_sq = scl.majority_prefers_2d( 60.0,  60.0, sq[0], sq[1], voters)

print("Majority preference (5 voters):")
print(f"  (110, 110) beats SQ: {a_beats_sq}")
print(f"  ( 60,  60) beats SQ: {b_beats_sq}")

pairwise_mat = scl.pairwise_matrix_2d(alts, voters)
print("\nPairwise matrix (1=WIN, 0=TIE, -1=LOSS):")
print(pairwise_mat)

# ===========================================================================
# 4. GEOMETRY: COPELAND, CONDORCET, UNCOVERED SET
# ===========================================================================

print("\n=== 4. GEOMETRY: COPELAND & CONDORCET ===")

copeland = scl.copeland_scores_2d(alts, voters)
copeland_winner = scl.copeland_winner_2d(alts, voters)
print(f"Copeland scores: {copeland}")
print(f"Copeland winner (0-based): {copeland_winner}")

has_cw = scl.has_condorcet_winner_2d(alts, voters)
cw = scl.condorcet_winner_2d(alts, voters)
print(f"Has Condorcet winner: {has_cw}")
print(f"Condorcet winner (0-based): {cw}")

uncovered = scl.uncovered_set_2d(alts, voters)
print(f"Uncovered set (0-based indices): {uncovered}")

# ===========================================================================
# 5. WINSET: MAJORITY WINSET OF STATUS QUO
# ===========================================================================

print("\n=== 5. WINSET ===")

ws = scl.winset_2d(sq[0], sq[1], voters)

print(f"Winset of SQ at {sq}:")
print(f"  Empty: {ws.is_empty()}")

if not ws.is_empty():
    bbox = ws.bbox()
    print(f"  Bounding box: [{bbox['min_x']:.1f}, {bbox['max_x']:.1f}] "
          f"× [{bbox['min_y']:.1f}, {bbox['max_y']:.1f}]")
    print(f"  Contains (110, 80):   {ws.contains(110.0, 80.0)}")
    print(f"  Contains (150, 150):  {ws.contains(150.0, 150.0)}")

# ===========================================================================
# 6. ORDINAL PROFILES & VOTING RULES
# ===========================================================================

print("\n=== 6. ORDINAL PROFILES & VOTING RULES ===")

profile = scl.profile_build_spatial(alts, voters)
dims = profile.dims()
print(f"Profile dimensions: {dims[0]} voters, {dims[1]} alternatives")

print("\nRankings (0-based indices, rank 0 = best):")
print(profile.export())

# Plurality
plurality = scl.plurality_scores(profile)
plurality_winner = scl.plurality_one_winner(profile)
print(f"\nPlurality scores: {plurality}")
print(f"Plurality winner (0-based): {plurality_winner}")

# Borda
borda = scl.borda_scores(profile)
borda_winner = scl.borda_one_winner(profile)
borda_ranking = scl.borda_ranking(profile)
print(f"\nBorda scores: {borda}")
print(f"Borda winner (0-based): {borda_winner}")
print(f"Borda ranking (best first): {borda_ranking}")

# Anti-plurality
antiplurality = scl.antiplurality_scores(profile)
antiplurality_winner = scl.antiplurality_one_winner(profile)
print(f"\nAnti-plurality scores: {antiplurality}")
print(f"Anti-plurality winner (0-based): {antiplurality_winner}")

# Approval voting (top-k), capped at n_alts
k_approval = min(2, dims[1])
approval_topk = scl.approval_scores_topk(profile, k=k_approval)
approval_topk_winners = scl.approval_all_winners_topk(profile, k=k_approval)
print(f"\nApproval voting (top-{k_approval}):")
print(f"  Scores:  {approval_topk}")
print(f"  Winners (0-based): {approval_topk_winners}")

# ===========================================================================
# 7. SOCIAL RANKINGS & PROPERTIES
# ===========================================================================

print("\n=== 7. SOCIAL RANKINGS & PROPERTIES ===")

scores = [1.5, 0.8, 2.1]
ranking = scl.rank_by_scores(scores)
print(f"Rank by scores {scores}:")
print(f"  Best-to-worst (0-based): {ranking}")

pareto = scl.pareto_set(profile)
print(f"\nPareto set (0-based): {pareto}")

for i in range(dims[1]):
    is_p = scl.is_pareto_efficient(profile, i)
    print(f"  Alt {i} is Pareto-efficient: {is_p}")

has_cw_prof = scl.has_condorcet_winner_profile(profile)
cw_prof = scl.condorcet_winner_profile(profile)
print(f"\nProfile has Condorcet winner: {has_cw_prof}")
print(f"Condorcet winner (profile, 0-based): {cw_prof}")

if cw_prof is not None:
    is_maj = scl.is_majority_consistent(profile, cw_prof)
    is_con = scl.is_condorcet_consistent(profile, cw_prof)
    print(f"  Is majority-consistent:  {is_maj}")
    print(f"  Is Condorcet-consistent: {is_con}")

# ===========================================================================
# 8. STREAM MANAGER & REPRODUCIBILITY
# ===========================================================================

print("\n=== 8. STREAM MANAGER & REPRODUCIBILITY ===")

sm = scl.StreamManager(master_seed=42, stream_names=["draws", "ties"])

draw1 = sm.uniform_real("draws", 0.0, 1.0)
draw2 = sm.uniform_real("draws", 0.0, 1.0)
print(f"Draws from stream 'draws':")
print(f"  Draw 1: {draw1}")
print(f"  Draw 2: {draw2}")

sm.reset_all(42)
draw1_again = sm.uniform_real("draws", 0.0, 1.0)
draw2_again = sm.uniform_real("draws", 0.0, 1.0)
print(f"After reset (master_seed=42):")
print(f"  Draw 1: {draw1_again}  (equal: {draw1 == draw1_again})")
print(f"  Draw 2: {draw2_again}  (equal: {draw2 == draw2_again})")

sm_ic = scl.StreamManager(master_seed=99, stream_names=["ic"])
profile_ic = scl.profile_impartial_culture(
    n_voters=10, n_alts=4, stream_manager=sm_ic, stream_name="ic"
)
ic_dims = profile_ic.dims()
print(f"\nImpartial culture profile (10 voters, 4 alts):")
print(f"  Dims: {ic_dims[0]} × {ic_dims[1]}")
print(f"  Voter 0 ranking: {profile_ic.get_ranking(0)}")

# ===========================================================================
# 9. CONFIGURATION FUNCTIONS
# ===========================================================================

print("\n=== 9. CONFIGURATION ===")

dist_euc      = scl.make_dist_config("euclidean", salience_weights=[1.0, 1.0])
dist_l1       = scl.make_dist_config("manhattan")
dist_weighted = scl.make_dist_config("euclidean", salience_weights=[2.0, 0.5])

print("Distance configs:")
print(f"  Euclidean (equal weights): {dist_euc}")
print(f"  Manhattan:                 {dist_l1}")
print(f"  Weighted euclidean:        {dist_weighted}")

loss_linear   = scl.make_loss_config("linear")
loss_quad     = scl.make_loss_config("quadratic")
loss_gaussian = scl.make_loss_config("gaussian", max_loss=2.0, steepness=0.5)

print("\nLoss configs:")
print(f"  Linear:    {loss_linear}")
print(f"  Quadratic: {loss_quad}")
print(f"  Gaussian (max_loss=2, steepness=0.5): {loss_gaussian}")

# ===========================================================================
# 10. API VERSION
# ===========================================================================

print("\n=== 10. API VERSION ===")

version = scl.scs_version()
print(f"C API version: {version['major']}.{version['minor']}.{version['patch']}")

# ===========================================================================
# 11. VISUALIZATION (canvas HTML — no plotly)
# ===========================================================================

print("\n=== 11. VISUALIZATION ===")

import socialchoicelab.plots as sclp

hull = scl.convex_hull_2d(voters)
bnd  = scl.uncovered_set_boundary_2d(voters, grid_resolution=10)

# Base plot: players + status quo only
fig = sclp.plot_spatial_voting(
    voters,
    sq          = sq,
    voter_names = scenario["voter_names"],
    dim_names   = tuple(scenario["dim_names"]),
    title       = scenario["name"],
)

fig = sclp.layer_winset(fig, ws)
fig = sclp.layer_convex_hull(fig, hull)
fig = sclp.layer_uncovered_set(fig, bnd)

_fd, _p = tempfile.mkstemp(suffix=".html")
os.close(_fd)
_tmp = Path(_p)
sclp.save_plot(fig, str(_tmp))
webbrowser.open(_tmp.as_uri())
print("Plot opened in browser. Hover over points for details.")

# ===========================================================================
# END OF SCRIPT
# ===========================================================================

print("\n=== END ===")
print("All examples completed successfully!")
print("For more detail, see: help(scl.function_name) or the API docs.")
